src/ncp/ncp_host.cpp

```diff
@@ -71,6 +71,71 @@ void NcpHost::GetDeviceRole(DeviceRoleHandler aHandler)
     mNcpSpinel.GetDeviceRole(aHandler);
 }
 
+void NcpHost::Join(const otOperationalDatasetTlvs &aActiveOpDatasetTlvs, const AsyncResultReceiver aReceiver)
+{
+    bool isAlready = !mJoinReceivers.empty();
+
+    auto handle_3 = [this](otError aError) {
+        CallAsyncReceivers(mJoinReceivers, aError,
+                           aError == OT_ERROR_NONE ? "Success" : "Failed to enable the Thread network!");
+    };
+
+    auto handle_2 = [this, handle_3](otError aError) {
+        if (aError != OT_ERROR_NONE)
+        {
+            CallAsyncReceivers(mJoinReceivers, aError, "Failed to enable the network interface!");
+        }
+        else
+        {
+            mNcpSpinel.ThreadSetEnabled(true, handle_3);
+        }
+    };
+
+    auto handle_1 = [this, handle_2](otError aError) {
+        if (aError != OT_ERROR_NONE)
+        {
+            CallAsyncReceivers(mJoinReceivers, aError, "Failed to set active dataset!");
+        }
+        else
+        {
+            mNcpSpinel.Ip6SetEnabled(true, handle_2);
+        }
+    };
```

Comment:
I think you can declare a class `AsyncResultReceiver` and let it store what the next receiver is. When a receiver is notified, you add the next action's receiver into the receiver list. In this way you can make chained actions.

Reply:
This is obviously not manageable. Can the NcpSpinel APIs be blocking?

Reply:
@wgtdkp 

> This is obviously not manageable. Can the NcpSpinel APIs be blocking?

Per our initial discussion, we decided to make all spinel operations for NCP as non-blocking. Because the blocking operations block the whole program, which we don't want to see.

We have an alternative: Add new spinel commands which bundle these operations together. For example, a command for 'join' is sent to NCP. And NCP will set active dataset, enable IP6, start thread and then respond to the host. The downside is that we need to add new spinel command. And if we don't update NCP, the behavior is hard to change. But I think this is still an option as we planned to make NCP updateable.

What do you mean by 'not manageable'? I prefer that we first find some better way to interpret the async process into code.

Reply:
I've updated to use a 'Then' style to implement the `Join` method. I introduced a utilitiy class `ChainedAsyncResult` which is basically as @superwhd mentioned.

The code is like:
```
    ChainedAsyncResultPtr result;
    bool                  isAlready = !mJoinReceivers.empty();

    VerifyOrExit(!isAlready);

    result = std::make_shared<ChainedAsyncResult>();
    result
        ->Then([this](otError aError, ChainedAsyncResultPtr aNext) {
            if (aError != OT_ERROR_NONE)
            {
                CallAsyncReceivers(mJoinReceivers, aError, "Failed to set active dataset!");
            }
            else
            {
                mNcpSpinel.Ip6SetEnabled(true, aNext);
            }
        })
        ->Then([this](otError aError, ChainedAsyncResultPtr aNext) {
            if (aError != OT_ERROR_NONE)
            {
                CallAsyncReceivers(mJoinReceivers, aError, "Failed to enable the network interface!");
            }
            else
            {
                mNcpSpinel.ThreadSetEnabled(true, aNext);
            }
        })
        ->Then([this](otError aError, ChainedAsyncResultPtr aNext) {
            OTBR_UNUSED_VARIABLE(aNext);
            CallAsyncReceivers(mJoinReceivers, aError,
                               aError == OT_ERROR_NONE ? "Success" : "Failed to enable the Thread network!");
        });

exit:
    mJoinReceivers.push_back(aReceiver);
    if (!isAlready)
    {
        mNcpSpinel.DatasetSetActiveTlvs(aActiveOpDatasetTlvs, result);
    }
```
What do you think?

Reply:
> Per our initial discussion, we decided to make all spinel operations for NCP as non-blocking. Because the blocking operations block the whole program, which we don't want to see.

Current OpenThread code is already implemented with blocking radio RPCs, isn't it?

Reply:
> Current OpenThread code is already implemented with blocking radio RPCs, isn't it?
Yes, but it's actually not a desired model. (It uses two `select`) The intial detailed discussion is [here](https://github.com/openthread/openthread/pull/9898#issuecomment-1984699024).

For RCP the blocking model can work because the radio operations are all small and quick to finish. When it comes to NCP, the operations can have much larger delay. For example, `DetachGracefully`. And a blocking model means the whole program will block on that operation. Then all the components in otbr-agent, including the platform things like netif, infrastructure if will all be blocked. So our idea is that, since we start NCP host from scratch, we use non-blocking model.

And the ThreadHost APIs are also async. It seems that the major concern is that the implementation code is not very readable. I think that with introduction of the utility class and the `Then` style operations, it's acceptable. Thoughts?

Reply:
The new result->Then()->Then() style is not readable neither, can we avoid it if possible?

> For RCP the blocking model can work because the radio operations are all small and quick to finish. When it comes to NCP, the operations can have much larger delay. For example, DetachGracefully. 

Can we make "setThreadEnabled()" blocking and make "DetachGracefully()" async? DetachGracefully() sends traffic to a peer device and waits for response, so it makes sense for it to be async.

Reply:
My concerns are around readability and maintenance. If there is a way to shorten this code to around three lines (because the blocking calls need only three lines), it will be great!

Reply:
@abtink @jwhui 

Do you have any preference on the implementation of the `Join` here? There are 3 options now:
1. Use the 'then' style chained async operations. (The current implementation). The major concern is the code readibility.
2. Adds a block mechanism for `NcpSpinel` and make the three operations (`SetDatasetActive`, `Ip6SetEnabled`, `ThreadStart`) as blocking methods. Will use the same approach as `RadioSpinel`.
3. Add a new spinel command which combines the 3 operations and let the NCP handle the new command.

Or do you have any other suggestions?

Reply:
Thanks guys.

I don't have any strong preferences or suggestions. Each model has its advantages and disadvantages. The `then` model seems like a good option.

Just to add another idea to the mix, we could also consider using co-routines (or classes that mimic co-routine behavior). `wpantund` used a co-routine-style mechanism called proto-thread, but I wouldn't necessarily recommend it here due to its potential complexity.

Reply:
Hostly coroutine is something new to me so I learned some basic things of it. And I think coroutine should be the ultimate solution for our case. 

As I understand, if we use coroutine, it will work like: we make the async ThreadHost APIs all as coroutines. And when it executes to spinel action, the coroutine is suspended and the flow returns to the caller of the Thread Host API. And then the code can continue to execute. And when a spinel response is obtained, the coroutine is resumed and execute again. This seems to be best solution that can give consideration to both performance and code readabilty maintainence.

The problem is how to use coroutine in `ot-br-posix`. The native c++ coroutine is available in c++20 and so cannot be used in `ot-br-posix` now. I noticed `boost` also has [coroutine2](https://www.boost.org/doc/libs/1_85_0/libs/coroutine2/doc/html/coroutine2/intro.html) and can be used with c++11. And boost is actually already used in `ot-br-posix` (used by `otbr-web`). So I think this is the most possbile choice. `proto-thread` may also be considered.

But anyway I hope we can first move forward with the current implementation to make otbr-agent can run with NCP. And since these are implementations (and no changes on NCP), we update and change them anytime. I'll follow up to seek a coroutine solution. Thoughts? @wgtdkp

Reply:
I am not okay with the fact that you need to write around 30 lines to just implement what can be implemented with three lines:

```
SuccessOrExit(error = setDataset());
SuccessOrExit(error = setThreadEnabled());
SuccessOrExit(error = setInterfaceUp());
```

Can you try boost coroutine and see how it works?

Reply:
Thanks! @wgtdkp 

I've further shorten the implementation. Now it's less than 10 lines:
```
    result = std::make_shared<ChainedAsyncResult>(errorHandler);
    result
        ->First([this, &aActiveOpDatasetTlvs](ChainedAsyncResultPtr aNext) {
            mNcpSpinel.DatasetSetActiveTlvs(aActiveOpDatasetTlvs, aNext);
        })
        ->Then([this](ChainedAsyncResultPtr aNext) { mNcpSpinel.Ip6SetEnabled(true, aNext); })
        ->Then([this](ChainedAsyncResultPtr aNext) { mNcpSpinel.ThreadSetEnabled(true, aNext); });
    result->Run();
```

---

I also tried to use `boost coroutine2`. It can be easily added to the project and build. However there are some difficulties to embedded it with our main loop.

If we use coroutine, the flow will be like:
1. Code executes a spinel operation, this starts a new 'context' and the action is suspend.
2. Then switch to the **main context** to process various events and tasks.

However in our current model, the call to spinel operation itself is from the main context. And there is no way to suspend the current main context while continuing to execute the main context.

In the introduction of boost.coroutine, it says
> Boost.Asio provides with its new asynchronous result feature a new framework combining event-driven model and coroutines, hiding the complexity of event-driven programming and permitting the style of classic sequential code. 

And gives an example:
```
void session(boost::asio::io_service& io_service){
    // construct TCP-socket from io_service
    boost::asio::ip::tcp::socket socket(io_service);

    try{
        for(;;){
            // local data-buffer
            char data[max_length];

            boost::system::error_code ec;

            // read asynchronous data from socket
            // execution context will be suspended until
            // some bytes are read from socket
            std::size_t length=socket.async_read_some(
                    boost::asio::buffer(data),
                    boost::asio::yield[ec]);
            if (ec==boost::asio::error::eof)
                break; //connection closed cleanly by peer
            else if(ec)
                throw boost::system::system_error(ec); //some other error

            // write some bytes asynchronously
            boost::asio::async_write(
                    socket,
                    boost::asio::buffer(data,length),
                    boost::asio::yield[ec]);
            if (ec==boost::asio::error::eof)
                break; //connection closed cleanly by peer
            else if(ec)
                throw boost::system::system_error(ec); //some other error
        }
    } catch(std::exception const& e){
        std::cerr<<"Exception: "<<e.what()<<"\n";
    }
}
```
I think in this example, with asio::yield, it transfer the flow to the main context which runs `io_service`. So if we want to use in that way we need to replace the current mainloop with asio. So I don't think it's worthwhile to do it now.